icu4c/source/i18n/unitsdata.cpp

```diff
@@ -315,37 +308,37 @@ int32_t getPreferenceMetadataIndex(const MaybeStackVector<UnitPreferenceMetadata
                                    UErrorCode &status) {
     if (U_FAILURE(status)) { return -1; }
     bool foundCategory, foundUsage, foundRegion;
-    int32_t idx = binarySearch(metadata, category, usage, region, &foundCategory, &foundUsage,
-                               &foundRegion, status);
+    UnitPreferenceMetadata desired(category, usage, region, -1, -1, status);
+    int32_t idx = binarySearch(metadata, desired, &foundCategory, &foundUsage, &foundRegion, status);
     if (U_FAILURE(status)) { return -1; }
     if (idx >= 0) { return idx; }
     if (!foundCategory) {
         status = U_ILLEGAL_ARGUMENT_ERROR;
-        return idx;
+        return -1;
     }
     U_ASSERT(foundCategory);
-    if (!foundUsage) {
-        if (uprv_strcmp(usage, "default") != 0) {
-            usage = "default";
-            idx = binarySearch(metadata, category, usage, region, &foundCategory, &foundUsage,
-                                       &foundRegion, status);
-        }
-        if (!foundUsage) {
+    while (!foundUsage) {
+        int32_t lastDashIdx = desired.usage.lastIndexOf('-');
+        if (lastDashIdx > 0) {
+            desired.usage.truncate(lastDashIdx);
+        } else if (uprv_strcmp(desired.usage.data(), "default") != 0) {
+            desired.usage.truncate(0).append("default", status);
+        } else {
             status = U_MISSING_RESOURCE_ERROR;
-            return idx;
+            return -1;
         }
+        idx = binarySearch(metadata, desired, &foundCategory, &foundUsage, &foundRegion, status);
```

Comment:
Check for the error code after this line, please.

Reply:
Done.

Reply:
(Still awaiting Travis results...)